src/riscv/linux/init.c

```diff
@@ -577,6 +658,41 @@ void cpuinfo_riscv_linux_init(void) {
 		linux_cpu_to_uarch_index_map[linux_id] = valid_uarchs_index - 1;
 	}
 
+	if (l2_count != 0) {
+		l2 = calloc(l2_count, sizeof(struct cpuinfo_cache));
+		if (l2 == NULL) {
+			cpuinfo_log_error("failed to allocate %zu bytes for descriptions of %zu L2 caches",
+							  l2_count * sizeof(struct cpuinfo_cache), l2_count);
+			goto cleanup;
+		}
+	}
+
+	/* Populate cache information structures in l2 */
+    uint32_t l2_index = UINT32_MAX;
+    for (uint32_t processor = 0; processor < valid_processors_count; processor++) {
+        struct cpuinfo_cache dummy_l1i, dummy_l1d, temp_l2 = { 0 };
+        cpuinfo_riscv_decode_cache(
+                riscv_linux_processors[processor].core.uarch,
+                &dummy_l1i, &dummy_l1d, &temp_l2);
+
+        if (temp_l2.size != 0) {
+            if (riscv_linux_processors[processor].package_leader_id == riscv_linux_processors[processor].processor.linux_id) {
+                l2_index += 1;
+                l2[l2_index] = (struct cpuinfo_cache) {
+                        .size            = temp_l2.size,
+                        .associativity   = temp_l2.associativity,
+                        .sets            = temp_l2.sets,
+                        .partitions      = temp_l2.partitions,
+                        .line_size       = temp_l2.line_size,
+                        .flags           = temp_l2.flags,
+                        .processor_start = processor,
+                        .processor_count = riscv_linux_processors[processor].package.processor_count,
+                };
+            }
+            processors[processor].cache.l2 = l2 + l2_index;
+        }
+    }
```

Comment:
Couldn't we do this in the loop above, right where we commit the L1 information?

Reply:
No because we don't know the value of `l2_count`. THe program first computes how many `l2` caches are present, then allocates the `l2` data structure and populates it. We can change the logic but it will behave diffferently from what it does now.